### ISSUE TITLE
BBC: Blead Breaks CPANEL/cPanel-TaskQueue-0.902.tar.gz

### ISSUE THREAD

**Issue by eserte** [Perl/perl5#19532]
t/taskqueue_pausing.t in CPANEL/cPanel-TaskQueue-0.902.tar.gz started to fail with perl 5.35.9:
```
#   Failed test 'Resume in other copy: TaskQueue is unpaused.'
#   at t/taskqueue_pausing.t line 46.
# Looks like you failed 1 test of 15.
t/taskqueue_pausing.t ....................... 
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/15 subtests 
```

First report at CPAN Testers: http://www.cpantesters.org/cpan/report/d8940380-99f0-11ec-81cf-263c376c1ca9
Overview of test reports: http://matrix.cpantesters.org/?dist=cPanel-TaskQueue%200.902

**Comment by jkeenan** [comment]
Bisecting with this invocation ...
```
perl Porting/bisect.pl \
--start=a6a6eee6da857dbb113d35fc617ecf81278fac64 \
--end=a4f7faaf7e7c49fceb79725d6853abbd83bc8bd9 \
--module=cPanel::TaskQueue
```
... points to the following commit:
```
59cc2f5b2548d81765e9f3dc2ce10c1070e787b8 is the first bad commit
commit 59cc2f5b2548d81765e9f3dc2ce10c1070e787b8
Author: Nicholas Clark <nick@ccl4.org>
Date:   Sun Jul 4 19:23:05 2021 +0000
Commit:     Nicholas Clark <nick@ccl4.org>
CommitDate: Sat Feb 19 19:46:37 2022 +0000


    Copy values that are "written as IV, then read as PV" with the same flags.
    
    Previously Perl_sv_setsv_flags() would gleefully turn on SVf_POK for these
    values, which meant that any copy no longer propagated the (new) state
    that says "this value started as an integer".

 sv.c | 6 ++++++
 1 file changed, 6 insertions(+)
bisect run success
That took 3761 seconds.
```
Sample test failure:
```
$ bleadprove -vb t/taskqueue_pausing.t
t/taskqueue_pausing.t .. 
1..15
ok 1 - TaskQueue is created unpaused.
ok 2 - TaskQueue is paused.
...
ok 12 - Resume in other copy: TaskQueue is paused.
ok 13 - Resume in other copy: TaskQueue2 is paused.
ok 14 - Resume in other copy: TaskQueue2 is unpaused.
not ok 15 - Resume in other copy: TaskQueue is unpaused.

\#   Failed test 'Resume in other copy: TaskQueue is unpaused.'
\#   at t/taskqueue_pausing.t line 46.
\# Looks like you failed 1 test of 15.
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/15 subtests 
```
@nwc10, can you take a look?

**Comment by nwc10** [comment]
That test file passes reliably for me with this hack:

```
diff --git a/lib/cPanel/StateFile.pm b/lib/cPanel/StateFile.pm
index dd8f235..1fd6738 100644
--- a/lib/cPanel/StateFile.pm
+++ b/lib/cPanel/StateFile.pm
@@ -300,6 +300,7 @@ sub import {
               or $state_file->throw("Unable to truncate the state file: $!");
 
             $state_file->{data_object}->save_to_cache( $state_file->{file_handle} );
+           sleep 3;
             $state_file->{file_mtime} = ( stat( $state_file->{file_handle} ) )[9];
 
             # Make certain we are at end of file.
```

Obviously that isn't really a fix, and it happens to cause several other tests in the distribution to fail.

I think the change in blead has exposed an existing subtle bug in the module. I find this code suspicious:

```
    sub _resynch {
        my ( $self, $guard, $mtime, $size ) = @_;

        if ( !$mtime || !$size ) {
            ( $mtime, $size ) = ( stat( $self->{file_name} ) )[ 9, 7 ];
        }

        # CPANEL-11795: Timewarp safety.  If time moves backwards we can loop forever
        if ( $self->{file_mtime} < $mtime || $self->{file_size} != $size || $self->{file_mtime} > time() ) {

            # File is newer or a different size
            $guard ||= cPanel::StateFile::Guard->new( { state => $self } );
            $guard->_open('+<');
            $self->{data_object}->load_from_cache( $self->{file_handle} );
            ( $self->{file_mtime}, $self->{file_size} ) = ( stat( $self->{file_handle} ) )[ 9, 7 ];
        }

        return $guard;
    }
```

I think that what is going wrong is that it **used** to be that the size of the serealised object changed (as a side effect of flags changing as a side effect of reading an integer in a string context) and that was sufficient to cause `$self->{file_size} != $size` to be false, and the reload to happen.

But now there is no size change, so the file size doesn't change, so the changed content isn't noticed, so no reload happens (immediately, which is what the test assumed). But I can't spot how to fix this problem without breaking other regression tests.

Basically, if I can "fix" a bug by adding `sleep` then the code in question has race conditions and needs fixing.

**Comment by toddr** [comment]
Thanks for letting us know. @jkeenan please do not consider this a blocker for the next version of Perl

**Comment by atoomic** [comment]
Thanks @jkeenan for bisecting and finding the commit which exposes the issue.

After investigation this is exposing a bug in `cPanel::StateFile::_resynch` itself which relies on a file timestamp.

As @nwc10 just pointed it this is a race condition based on time...

The fact this is happening after the change from https://github.com/cpanel/cPanel-TaskQueue/blob/55494c86d45a671d3ec76d9fb7b8ac591ad70800/lib/cPanel/StateFile.pm#L462
is surprising but I think it can be explained by code running faster than previously as some OPs are not necessary anymore for some PVs... in some case

I will provide a cleaner fix to the cPanel::TaskQueue repo.

But here is for a dirty workaround which solves the issue discovered here without adding a sleep
There is nothing wrong to Perl here. Just a bad design upstream.

```diff
diff --git a/lib/cPanel/StateFile.pm b/lib/cPanel/StateFile.pm
index dd8f235..14c50bf 100644
--- a/lib/cPanel/StateFile.pm
+++ b/lib/cPanel/StateFile.pm
@@ -459,7 +459,7 @@ sub import {
         }

         # CPANEL-11795: Timewarp safety.  If time moves backwards we can loop forever
-        if ( $self->{file_mtime} < $mtime || $self->{file_size} != $size || $self->{file_mtime} > time() ) {
+        if ( 1 ) {

             # File is newer or a different size
             $guard ||= cPanel::StateFile::Guard->new( { state => $self } );
```

We should close this case, thanks @eserte  for reporting 

I think the lesson learnt here is that code sensible to time race condition are going to be impacted by performance optimization (or slow down too).

Thanks

**Comment by atoomic** [comment]
FYI I released cPanel-TaskQueue v0.903 to fix the issue discovered here.

https://metacpan.org/release/CPANEL/cPanel-TaskQueue-0.903/view/lib/cPanel/TaskQueue.pm

Thank you everyone for your contribution
